plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -1359,23 +1361,70 @@ public void setColumnType(ConnectorSession session, ConnectorTableHandle tableHa
+        verifyStorageFormatForColumnTypeChange(storageFormat);
+        verifySupportedColumnTypeChange(hiveColumn.getType(), type);
+        if (table.getPartitionColumns().stream().anyMatch(partition -> partition.getName().equals(hiveColumn.getName()))) {
+            throw new TrinoException(NOT_SUPPORTED, "Changing partition column types is not supported");
+        }
+        ImmutableList.Builder<Partition> partitionsBuilder = ImmutableList.builder();
+        List<String> partitionsNames = metastore.getPartitionNames(table.getSchemaName(), table.getTableName())
+                .orElseThrow(() -> new TableNotFoundException(table.getSchemaTableName()));
+        for (List<String> partitionsNamesBatch : Lists.partition(partitionsNames, 1000)) {
+            metastore.getPartitionsByNames(table.getSchemaName(), table.getTableName(), partitionsNamesBatch).values().stream()
+                    .filter(Optional::isPresent).map(Optional::get)
+                    .forEach(partition -> {
+                        boolean skipUpdatePartition = false;
+                        verifyStorageFormatForColumnTypeChange(extractHiveStorageFormat(partition.getStorage().getStorageFormat()));
+                        ImmutableList.Builder<Column> columns = ImmutableList.builder();
+                        for (Column column : partition.getColumns()) {
+                            if (column.getName().equals(hiveColumn.getName())) {
+                                Type sourceType = column.getType().getType(typeManager, getTimestampPrecision(session));
+                                verifySupportedColumnTypeChange(sourceType, type);
+                                columns.add(column);
+                                skipUpdatePartition = sourceType.equals(type);
+                            }
+                            else {
+                                columns.add(column);
+                            }
+                        }
+                        if (!skipUpdatePartition) {
+                            partitionsBuilder.add(Partition.builder(partition).setColumns(columns.build()).build());
+                        }
+                    });
+        }
+        List<Partition> partitions = partitionsBuilder.build();
+        metastore.setColumnType(table.getSchemaName(), table.getTableName(), hiveColumn.getName(), toHiveType(type));
+        if (!partitions.isEmpty()) {
+            // Hive changes a column definition in each partitions unless the ALTER TABLE statement doesn't contain partition condition
+            // Trino doesn't support specifying partitions in ALTER TABLE, so SET DATA TYPE updates all partitions
+            // https://cwiki.apache.org/confluence/display/hive/languagemanual+ddl#LanguageManualDDL-AlterPartition
+            metastore.alterPartitions(table.getSchemaName(), table.getTableName(), partitions, OptionalLong.empty());
```

Comment:
I am surprised 

Firstl, i wouldn't expect existing partitions to be updated at all.

Second, it will make it impossible to update type in large tables.

Let's talk more about this